### PR TITLE
DR-995 add database restrictions to protect against conflicting concurrent calls.

### PR DESCRIPTION
**Pull request by myessail**
Adds a constraint on the `asset_specification` table that doesn't allow more than one asset with the same name per dataset. This was previously being checked in `CreateDatasetAssetStep`, but we could get into a state where two threads could pass the check before committing their changes to the database. By making that change transactional and enforcing the restriction above, we can guarantee that this will now never happen.

### CODE REVIEW

**Review by fboulnois — APPROVED** [review]
lgtm

File: src/main/java/bio/terra/service/dataset/AssetDao.java [hunk 1]
```diff
@@ -34,6 +38,14 @@ public AssetDao(DataRepoJdbcConfiguration jdbcConfiguration) {
                 .collect(Collectors.toList());
     }
 
+    /**
+     * Create a new AssetSpecification. If a DuplicateKeyException is thrown,
+     * catch it and throw an InvalidAssetException
+     * @param assetSpecification
```
Comment (ddietterich) [inline src/main/java/bio/terra/service/dataset/AssetDao.java]: nit: worth filling in the javadoc?

File: src/main/java/bio/terra/service/dataset/flight/create/CreateDatasetAssetStep.java [hunk 2]
```diff
@@ -92,24 +90,35 @@ public StepResult doStep(FlightContext context) {
         } catch (Exception e) {
```
Comment (ddietterich) [inline src/main/java/bio/terra/service/dataset/flight/create/CreateDatasetAssetStep.java:90]: Not your code, but I'm curious why the fault is surrounded by a try catch. It catches its own exception and then retrhows it. Seems odd.

File: src/main/resources/db/changesets/20200529_datasetuniqueasset.yaml [hunk 3]
```diff
@@ -0,0 +1,14 @@
+databaseChangeLog:
+  - changeSet:
+      id: datasetUniqueAsset
+      author: myessail
+      changes:
+#        - sql:
```
Comment (ddietterich) [inline src/main/resources/db/changesets/20200529_datasetuniqueasset.yaml]: Remove dead code?

**Review by ddietterich — APPROVED** [review]
LGTM - a few nits

File: src/main/java/bio/terra/service/dataset/AssetDao.java [hunk 4]
```diff
@@ -34,6 +38,14 @@ public AssetDao(DataRepoJdbcConfiguration jdbcConfiguration) {
                 .collect(Collectors.toList());
     }
 
+    /**
+     * Create a new AssetSpecification. If a DuplicateKeyException is thrown,
+     * catch it and throw an InvalidAssetException
```
Comment (marikomedlock) [inline src/main/java/bio/terra/service/dataset/AssetDao.java]: Nit: Maybe make this a little more reader-friendly? If you try to create an asset with the same name as an existing one for the same dataset, this method throws an InvalidAssetException. Or something like that. I think that might be a little better than talking about a DuplicateKeyException specifically.

File: src/test/java/bio/terra/service/dataset/DatasetServiceTest.java [hunk 5]
```diff
@@ -165,6 +166,120 @@ public void addDatasetAssetSpecifications() throws Exception {
...
+        String jobId2 = datasetService.addDatasetAssetSpecifications(datasetId2.toString(), assetModel, testUser);
+        flightIdsList.add(jobId2);
+
+        boolean assetAdd2 = TestUtils.eventualExpect(5, 60, true, () ->
+            jobService.retrieveJob(jobId2, testUser).getJobStatus().equals(JobModel.JobStatusEnum.SUCCEEDED)
+        );
+        Assert.assertTrue(assetAdd2);
+
+        Dataset dataset2 = datasetDao.retrieve(datasetId2);
+        // make sure the first asset we created hasn't been deleted during the undo step
```
Comment (marikomedlock) [inline src/test/java/bio/terra/service/dataset/DatasetServiceTest.java]: Nit: there should be no undo in this case, because both succeeded right?

**Review by marikomedlock — APPROVED** [review]
LGTM. Two minor comments about comments.

**Review by fboulnois — APPROVED** [review]
Latest tests lgtm

**Pull request merged**